src/execution.rs

```diff
@@ -112,7 +148,10 @@ impl File {
         for stanza in &self.stanzas {
             stanza.execute(
                 tree,
+                source_path,
                 source,
+                tsg_path,
+                tsg_source,
```

Comment:
I had a quick go at bundling up both path and source into a `struct`, but ultimately didn't feel it was worth the space.